### PR TITLE
[PyTorch] FP8 Tensor improvements

### PR DESCRIPTION
**Pull request by ksivaman**

### CODE REVIEW

File: transformer_engine/pytorch/module/linear.py [hunk 1]
```diff
@@ -314,9 +308,9 @@ def backward(
                 weight_t_fp8,
                 fwd_scale_inverses,
             ) = ctx.saved_tensors
-            if weight_t_fp8 is not None:
-                assert hasattr(weight_t_fp8, "_data"), \
-                       "_data attr doesn't exist (after restore in bwd)"
+
+            if ctx.fp8 and weight_t_fp8 is None:
+                weight_t_fp8 = weight.transpose(update_cache=ctx.is_first_microbatch)
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/module/linear.py:313]: If `ctx.fp8` is True and `weight_t_fp8` is `None`, it means that `primary_weights_in_fp8` is `True`

[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/module/linear.py:316]: Yeah I think it'd be a good idea to add this as a small comment

**Comment by sudhakarsingh27** [comment]
Lgtm! I was able to verify expected mem savings with `Linear` and `LayerNormLinear` (I don't have a model currently that could check `LayerNormMLP` right away)

**Review by timmoon10 — APPROVED** [review]

**Pull request merged**